go/libkb/secret_store_darwin.go

```diff
@@ -7,138 +7,226 @@ package libkb
+	item.SetAccessible(k.accessible())
+	return keychain.AddItem(item)
 }
 
-func (k KeychainSecretStore) updateAccessibility(m MetaContext, accountName string) {
+func (k KeychainSecretStore) updateAccessibility(mctx MetaContext, account keychainSlottedAccount) {
 	query := keychain.NewItem()
 	query.SetSecClass(keychain.SecClassGenericPassword)
-	query.SetService(k.serviceName(m))
-	query.SetAccount(accountName)
+	query.SetService(k.serviceName(mctx))
+	query.SetAccount(account.String())
 
 	// iOS keychain returns `keychain.ErrorParam` if this is set so we skip it.
 	if !isIOS {
 		query.SetMatchLimit(keychain.MatchLimitOne)
 	}
+
 	updateItem := keychain.NewItem()
 	updateItem.SetAccessible(k.accessible())
 	if err := keychain.UpdateItem(query, updateItem); err != nil {
-		m.Debug("KeychainSecretStore.updateAccessibility: failed: %s", err)
+		mctx.Debug("KeychainSecretStore.updateAccessibility: failed: %s", err)
 	}
 }
 
-func (k KeychainSecretStore) mobileKeychainPermissionDeniedCheck(m MetaContext, err error) {
-	m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: checking for mobile permission denied")
-	if !isIOS || m.G().GetAppType() != MobileAppType {
-		m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: not an iOS app")
+func (k KeychainSecretStore) mobileKeychainPermissionDeniedCheck(mctx MetaContext, err error) {
+	mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: checking for mobile permission denied")
+	if !isIOS || mctx.G().IsMobileAppType() {
+		mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: not an iOS app")
 		return
 	}
 	if err != keychain.ErrorInteractionNotAllowed {
-		m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: wrong kind of error: %s", err)
+		mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: wrong kind of error: %s", err)
 		return
 	}
-	m.G().Log.Warning("mobileKeychainPermissionDeniedCheck: keychain permission denied: %s", err)
+	mctx.G().Log.Warning("mobileKeychainPermissionDeniedCheck: keychain permission denied, aborting: %s", err)
 	os.Exit(4)
 }
 
-func (k KeychainSecretStore) RetrieveSecret(m MetaContext, accountName NormalizedUsername) (LKSecFullSecret, error) {
-	m.Debug("KeychainSecretStore.RetrieveSecret(%s)", accountName)
-	encodedSecret, err := keychain.GetGenericPassword(k.serviceName(m), string(accountName), "", "")
+func (k KeychainSecretStore) RetrieveSecret(mctx MetaContext, accountName NormalizedUsername) (secret LKSecFullSecret, err error) {
+	defer mctx.TraceTimed(fmt.Sprintf("KeychainSecretStore.RetrieveSecret(%s)", accountName), func() error { return err })()
+
+	// find the last valid item we have stored in the keychain
+	var previousSecret LKSecFullSecret
+Loop:
+	for i := 0; i < maxKeychainItemSlots; i++ {
+		account := newKeychainSlottedAccount(accountName, i)
+		secret, err = k.retrieveSecret(mctx, account)
+		switch err {
+		case nil:
+			previousSecret = secret
+		case keychain.ErrorItemNotFound, NewErrSecretForUserNotFound(accountName):
```

Comment:
`err` will never equal the object returned by NewErrSecretForUserNotFound, right?  you could check the types for equality.

Reply:
we get equality since the tests depend this error being in the switch but think the type check will be cleaner